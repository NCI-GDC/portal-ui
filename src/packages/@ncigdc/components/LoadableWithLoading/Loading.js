/* @flow */

import React from 'react';
import Overlay from '@ncigdc/uikit/Overlay';
import Spinner from '@ncigdc/uikit/Loaders/Material';

const Loading = ({
  isLoading,
  timedOut,
  pastDelay,
  error,
}: {
  isLoading: boolean,
  timedOut: boolean,
  pastDelay: boolean,
  error: boolean,
}) => {
  if (isLoading) {
    // While our other component is loading...
    if (timedOut) {
      // In case we've timed out loading our other component.
      return <div>Loader timed out!</div>;
    } else if (pastDelay) {
      // Display a loading screen after a set delay.
      return (
        <Overlay show={true} style={{ position: 'absolute', zIndex: 10 }}>
          <Spinner />
        </Overlay>
      );
    } else {
      // Don't flash "Loading..." when we don't need to.
      return null;
    }
  } else if (error) {
    // If we aren't loading, maybe
    return <div>Error! Component failed to load</div>;
  } else {
    // This case shouldn't happen... but we'll return null anyways.
    return null;
  }
};

export default Loading;
