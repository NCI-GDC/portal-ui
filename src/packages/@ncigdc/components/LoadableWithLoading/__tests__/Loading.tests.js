/* @flow */
import React from 'react';
import renderer from 'react-test-renderer';
import Loading from '../Loading';

describe('Loading', () => {
  it('handles isLoading before pastDelay', () => {
    const props = {
      isLoading: true,
      timedOut: false,
      pastDelay: false,
      error: false,
    };
    const component = renderer.create(<Loading {...props} />);
    expect(component).toMatchSnapshot(); //shows null
  });
  it('handles timedOut', () => {
    const props = {
      isLoading: true,
      timedOut: true,
      pastDelay: false,
      error: false,
    };
    const component = renderer.create(<Loading {...props} />);
    expect(component).toMatchSnapshot(); //shows timed out, note: currently timeout is not set, will keep waiting
  });
  it('handles pastDelay', () => {
    const props = {
      isLoading: true,
      timedOut: false,
      pastDelay: true,
      error: false,
    };
    const component = renderer.create(<Loading {...props} />);
    expect(component).toMatchSnapshot(); //shows spinner
  });
  it('handles error', () => {
    const props = {
      isLoading: false,
      timedOut: false,
      pastDelay: false,
      error: true,
    };
    const component = renderer.create(<Loading {...props} />);
    expect(component).toMatchSnapshot(); //shows error message
  });
});
