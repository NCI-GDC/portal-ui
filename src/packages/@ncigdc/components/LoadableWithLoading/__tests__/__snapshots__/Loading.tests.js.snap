// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading handles error 1`] = `
<div>
  Error! Component failed to load
</div>
`;

exports[`Loading handles isLoading before pastDelay 1`] = `null`;

exports[`Loading handles pastDelay 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgba(255, 255, 255, 0.8)",
      "cursor": "default",
      "display": "flex",
      "height": "100%",
      "justifyContent": "center",
      "left": 0,
      "opacity": 1,
      "pointerEvents": "all",
      "position": "absolute",
      "top": 0,
      "transition": "opacity 0.35s ease",
      "width": "100%",
      "zIndex": 10,
    }
  }
>
  <div
    className="mdl-spinner mdl-spinner--single-color mdl-js-spinner is-active is-upgraded"
    style={
      Object {
        "transition": "opacity 0.35s ease",
      }
    }
  >
    <div
      className="mdl-spinner__layer mdl-spinner__layer-1"
      style={
        Object {
          "borderColor": "rgb(66, 133, 244)",
        }
      }
    >
      <div
        className="mdl-spinner__circle-clipper mdl-spinner__left"
      >
        <div
          className="mdl-spinner__circle"
        />
      </div>
      <div
        className="mdl-spinner__gap-patch"
      >
        <div
          className="mdl-spinner__circle"
        />
      </div>
      <div
        className="mdl-spinner__circle-clipper mdl-spinner__right"
      >
        <div
          className="mdl-spinner__circle"
        />
      </div>
    </div>
    <div
      className="mdl-spinner__layer mdl-spinner__layer-2"
      style={
        Object {
          "borderColor": "#123e57",
        }
      }
    >
      <div
        className="mdl-spinner__circle-clipper mdl-spinner__left"
      >
        <div
          className="mdl-spinner__circle"
        />
      </div>
      <div
        className="mdl-spinner__gap-patch"
      >
        <div
          className="mdl-spinner__circle"
        />
      </div>
      <div
        className="mdl-spinner__circle-clipper mdl-spinner__right"
      >
        <div
          className="mdl-spinner__circle"
        />
      </div>
    </div>
    <div
      className="mdl-spinner__layer mdl-spinner__layer-3"
      style={
        Object {
          "borderColor": "#bb0e3d",
        }
      }
    >
      <div
        className="mdl-spinner__circle-clipper mdl-spinner__left"
      >
        <div
          className="mdl-spinner__circle"
        />
      </div>
      <div
        className="mdl-spinner__gap-patch"
      >
        <div
          className="mdl-spinner__circle"
        />
      </div>
      <div
        className="mdl-spinner__circle-clipper mdl-spinner__right"
      >
        <div
          className="mdl-spinner__circle"
        />
      </div>
    </div>
    <div
      className="mdl-spinner__layer mdl-spinner__layer-4"
      style={
        Object {
          "borderColor": "#5b5151",
        }
      }
    >
      <div
        className="mdl-spinner__circle-clipper mdl-spinner__left"
      >
        <div
          className="mdl-spinner__circle"
        />
      </div>
      <div
        className="mdl-spinner__gap-patch"
      >
        <div
          className="mdl-spinner__circle"
        />
      </div>
      <div
        className="mdl-spinner__circle-clipper mdl-spinner__right"
      >
        <div
          className="mdl-spinner__circle"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Loading handles timedOut 1`] = `
<div>
  Loader timed out!
</div>
`;
